File: reverse_string.c
// 344. Reverse String
#include "leetcode.h"

/*
 * write a function that reverses a string. the input string is given as an
 * array of characters 's'. you must do this by modifying teh input array in
 * place with 'O(1)' extra memroy
 */

void reverseString(char *s, int sSize) {
  char *ptr = s + sSize - 1, tmp = *s;
  if (sSize <= 1)
    return;
  *s = *ptr;
  *ptr = tmp;
  reverseString(s + 1, sSize - 2);
}

int main() {
  char *s1 = "hello", *s2 = "Hannah";
  reverseString(s1, strlen(s1));
  reverseString(s2, strlen(s2));
  printf("%s\n", s1); // expect: olleh
  printf("%s\n", s2); // expect: hannaH
}

File: reverse_string.py
# 344. Reverse String

"""
write a function that reverses a string. the input string is given as an
array of characters 's'. you must do this by modifying teh input array in
place with 'O(1)' extra memroy
"""


class Solution(object):
    def reverseString(self, s):
        """
        :type s: List[str]
        :rtype: None Do not return anything, modify s in-place instead.
        """
        s[:] = s[::-1]


if __name__ == "__main__":
    obj = Solution()
    s1 = "hello"
    s2 = "Hannah"
    obj.reverseString(s1)
    obj.reverseString(s2)
    print(s1)
    print(s2)
